zerver/actions/uploads.py

```diff
@@ -90,23 +90,27 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
 
 def check_attachment_reference_change(
     message: Union[Message, ScheduledMessage], rendering_result: MessageRenderingResult
-) -> bool:
+) -> Tuple[bool, List[Dict[str, Any]]]:
     # For a unsaved message edit (message.* has been updated, but not
     # saved to the database), adjusts Attachment data to correspond to
     # the new content.
+    # detached_attachments: List[Dict[str, Any]] = []
     prev_attachments = {a.path_id for a in message.attachment_set.all()}
     new_attachments = set(rendering_result.potential_attachment_path_ids)
 
     if new_attachments == prev_attachments:
-        return bool(prev_attachments)
-
+        return bool(prev_attachments), []
     to_remove = list(prev_attachments - new_attachments)
     if len(to_remove) > 0:
         attachments_to_update = Attachment.objects.filter(path_id__in=to_remove).select_for_update()
         message.attachment_set.remove(*attachments_to_update)
 
+    detached_attachments_query = Attachment.objects.filter(
+        path_id__in=to_remove, messages__isnull=True
+    )
+    detached_attachments = [d.to_dict() for d in detached_attachments_query]
```

Comment:
So I have a concern about this approach -- I believe that in some cases the current user will not be the owner of the upload, and not have permission to remove it.

I think this logic will need to extend the filter in this query. Please also write backend tests that verify the basic logic for this feature as well as this specific corner case (where another user uploads a file, sends a message linking to it, this user sends a second message linking to it, then the first user deletes their message and the second edits theirs to no longer link to it -- this popup should not occur in that case).